_posts/2021-05-06-broker-tuning.md

```diff
@@ -0,0 +1,441 @@
+
+Pay attention to the configuration requirements for internal Kafka topics.
+
+If you are [using transactions](https://strimzi.io/docs/operators/latest/using.html#reliability_guarantees) to guarantee the reliability of writes to partitions from producers, the internal `__transaction_state` topic used in the process requires a minimum of three brokers with the default settings.
+
+```properties
+# ...
+transaction.state.log.replication.factor=3
+transaction.state.log.min.isr=2
+# ...
+```
+
+The `__consumer_offsets` topic, which stores partition offset commits, has default settings for the number of partitions and replication factor.
+
+```properties
+# ...
+offsets.topic.num.partitions=50
+offsets.topic.replication.factor=3
+offsets.commit.required.acks=-1
+# ...
+```
+
+***:warning: Do not reduce these settings in production.***
+
+### Improving request handling throughput by increasing I/O threads
+
+Network threads (`num.network.threads`) handle requests to the Kafka cluster, such as produce and fetch requests from client applications.
+Adjust the number of network threads to reflect the replication factor and the levels of activity from client producers and consumers interacting with the Kafka cluster.
+
+I/O threads (`num.io.threads`) pick up requests from the request queue to process them. Adding more threads can improve throughput, but the number of CPU cores/and disk bandwidth imposes a practical upper limit.
+The minimum number of threads should equal the number of storage volumes.
+
+To reduce congestion and regulate the request traffic, you can use `queued.max.requests` to limit the number of requests allowed in the request queue before the network thread is blocked.
+Use `num.recovery.threads.per.data.dir` to specify the number of threads used for log loading at startup and flushing at shutdown.
+
+```properties
+# ...
+num.io.threads=8
+queued.max.requests=500
+num.network.threads=3
+num.recovery.threads.per.data.dir=1
+# ...
+```
+
+> Kafka broker metrics can help with working out the number of threads required. For example, metrics for the average time network threads are idle (`kafka.network:type=SocketServer,name=NetworkProcessorAvgIdlePercent`) indicate the percentage of resources used.
+If there is 0% idle time, all resources are in use, which means that more threads would be beneficial.
+
+If threads are slow or limited due to the number of disks, try increasing the size of the buffers for network requests to improve throughput:
+
+```properties
+# ...
+replica.socket.receive.buffer.bytes=65536
+# ...
+```
+
+And also increase the maximum number of bytes Kafka can receive:
+
+```properties
+# ...
+socket.request.max.bytes=104857600
+# ...
+```
+
+### Increasing bandwidth for high latency connections
+
+If you've fine-tuned the size of your message batches, but latency is impeding the performance of your Kafka environment, try increasing the size of the buffers for sending and receiving messages.
```

Comment:
Not sure I understand what is this trying to say. Increasing buffers normally improves throughput at the cost of higher latency. Is that the case here? Maybe we can make it a bit more clear.

Reply:
Changed to: 

> If you've fine-tuned the size of your message batches, the default values of the buffers for sending and receiving messages might be too small for the required throughput.